interfaces/snap_app_set.go

```diff
@@ -104,91 +106,83 @@ func (a *SnapAppSet) SecurityTagsForSlot(slot *snap.SlotInfo) ([]string, error)
-// constructed from the apps and hooks that are associated with the slot.
-func (a *SnapAppSet) SlotLabelExpression(slot *ConnectedSlot) string {
-	// TODO: this is a hack that will not continue to work once component hooks
-	// are introduced. the methods on SnapAppSet should only be called on
-	// slots/hooks that originated from the snap that the SnapAppSet was derived
-	// from.
-	info := a.info
-	if a.info.InstanceName() != slot.slotInfo.Snap.InstanceName() {
-		info = slot.slotInfo.Snap
-	}
-
-	apps := info.AppsForSlot(slot.slotInfo)
-	hooks := info.HooksForSlot(slot.slotInfo)
-	return labelExpr(apps, hooks, info)
+	return labelExpr(a, a.PlugRunnables(plug))
 }
 
-// RunnableType is an enumeration of the different types of runnables that can
-// be present in a snap.
-type RunnableType int
-
-const (
-	RunnableApp RunnableType = iota
-	RunnableHook
-	RunnableComponentHook
-)
+// slotLabelExpression returns the label expression for the given slot. It is
+// constructed from the apps and hooks that are associated with the slot.
+func (a *SnapAppSet) slotLabelExpression(slot *ConnectedSlot) string {
+	if a.info.InstanceName() != slot.appSet.InstanceName() {
+		panic("internal error: connected slot must be from the same snap as the SnapAppSet")
+	}
 
-// Runnable represents a runnable element of a snap.
-type Runnable struct {
-	// CommandName is the name of the command that is run when this runnable
-	// runs.
-	CommandName string
-	// SecurityTag is the security tag associated with the runnable. Security
-	// tags are used by various security subsystems as "profile names" and
-	// sometimes also as a part of the file name.
-	SecurityTag string
+	return labelExpr(a, a.SlotRunnables(slot))
 }
 
 // Runnables returns a list of all runnables known by the app set.
-func (a *SnapAppSet) Runnables() []Runnable {
-	var runnables []Runnable
+func (a *SnapAppSet) Runnables() []snap.Runnable {
+	var runnables []snap.Runnable
 
 	for _, app := range a.info.Apps {
-		runnables = append(runnables, Runnable{
-			CommandName: app.Name,
-			SecurityTag: app.SecurityTag(),
-		})
+		runnables = append(runnables, snap.AppRunnable(app))
 	}
 
 	for _, hook := range a.info.Hooks {
-		runnables = append(runnables, Runnable{
-			CommandName: fmt.Sprintf("hook.%s", hook.Name),
-			SecurityTag: hook.SecurityTag(),
-		})
+		runnables = append(runnables, snap.HookRunnable(hook))
 	}
 
 	for _, component := range a.components {
 		for _, hook := range component.Hooks {
-			runnables = append(runnables, Runnable{
-				CommandName: fmt.Sprintf("%s.hook.%s", component.Component, hook.Name),
-				SecurityTag: hook.SecurityTag(),
-			})
+			runnables = append(runnables, snap.HookRunnable(hook))
 		}
 	}
 
 	return runnables
 }
 
+// PlugRunnables returns a list of all runnables that should be connected to the
+// given plug.
+func (a *SnapAppSet) PlugRunnables(plug *ConnectedPlug) []snap.Runnable {
```

Comment:
now that ConnectedPlug has an SnapAppSet can't this be a ConnectedPlug.Runnables method?